### PR TITLE
Fix up matrix updates

### PR DESCRIPTION
**Pull request by gfodor**
This PR performs an internal overhaul and a few monkey patches to three.js to reduce the amount of matrix operations for local + world xforms in the steady state case by approx 80-90%.

- Disable the automatic matrix update loop on the scene and takes it over via an A-Frame system called `world-update`

- Turns off `matrixAutoUpdate` by default on all objects, and adds the `matrix-auto-update` component which can be added piecemeal to entities to turn it back on. Also turns it on for imported GLTF nodes that have animations.

- Introduces the `matrixNeedsUpdate` flag which will cause the local and world matrix to be updated on that object at the end of the frame.

- Adds `Object3D.updateMatrices` which will update the local and world matrix on an object and (optionally) all of its parents. (doing the parents first, so the final world matrices will be correct.) This should generally be what you call instead of `updateMatrixWorld(true)` if you want to forcibly update the world matrix.

- Updates `getWorldPosition`, etc. to no longer use `updateMatrixWorld` but now instead use `updateMatrices`

- Adds helper functions to get world space position/rotation/scale without performing a world matrix update.

- Re-implements `updateMatrixWorld` with two major optimizations. First, non-visible objects will no longer have their local nor world matrices updated (nor their children.) Second, objects which never have their local matrices changed from the identity matrix end up just copying by reference the world matrix from the parent during the world matrix update. (This is extremely common in A-Frame)

- Fixes one or two weird bugs in the GLTF loader that seemed to be paved over when we were updating matrices every frame.

The net result of whitelisting auto matrix updates, avoiding non-visible updates, and skipping world matrix updates for identity matrices results in a test scene with 5 ducks dropping from 384 local matrix updates and 388 world matrix updates to 18 and 38 respectively.

### CODE REVIEW

**Comment by gfodor** [comment]
some quick benchmarks of running a scene with 5 ducks, average ms per global world matrix update in Chrome desktop across 750 frames:

```
baseline: 0.388 0.407 0.438
switch to manual matrix update flag: 0.418 0.426 0.397
hold out visbility optimization: 0.252 0.283 0.247
hold out identity optimization: 0.264 0.259 0.260
full optimizations: 0.211 0.204 0.218
```

**Comment by gfodor** [comment]
this PR now includes a new component which will set all non-positional audio nodes to not be visible, so their matrices are skipped altogether

File: src/systems/world-update.js [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      }
+    };
+  },
+
+  _patchRenderFunc: function() {
+    const renderer = this.el.renderer;
+    const render = renderer.render;
+
+    renderer.render = (scene, camera, renderTarget) => {
+      scene.updateMatrixWorld();
```
Comment (netpro2k) [inline src/systems/world-update.js]: Hmm we set `scene.autoUpdate` to `false` but then we patch render to basically turn it back on. What am I missing?

File: src/systems/world-update.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      render.call(renderer, scene, camera, renderTarget);
+      this.frame++;
```
Comment (netpro2k) [inline src/systems/world-update.js]: We don't seem to use this anywhere. Just for debugging? The camera and any post effects are probably going to mess this count up.

File: src/systems/world-update.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    // Computes this object's matrices and then the recursively computes the matrices of all the children.
+    //
+    // forceWorldUpdate - If true and the object is visible, will force the world matrix to be updated for
+    // this node and all of its children.
+    //
+    // includeInvisible - If true, does not ignore non-visible objects.
+    THREE.Object3D.prototype.updateMatrixWorld = function(forceWorldUpdate, includeInvisible) {
+      if (!this.visible && !includeInvisible) return;
+
+      // Do not recurse upwards, since this is recursing downwards
```
Comment (netpro2k) [inline src/systems/world-update.js]: confused by this comment. `this.updateMatricies` only recurses upwards no? What is the "this" in "this is only recursing downwards"

File: src/systems/world-update.js [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    //
+    // forceWorldUpdate - Forces the world matrix to be updated regardless of if the local matrix
+    // has been updated since the last update.
+    //
+    // skipParents - unless true, all parent matricies are updated before updating this object's
+    // local and world matrix.
+    //
+    // Returns true if the world matrix was updated.
+    //
+    THREE.Object3D.prototype.updateMatrices = function(forceLocalUpdate, forceWorldUpdate, skipParents) {
```
Comment (netpro2k) [inline src/systems/world-update.js]: This is pretty dense so I would rather just talk through it over voice.

File: src/systems/world-update.js [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+        if (this.cachedMatrixWorld) {
+          this.cachedMatrixWorld.copy(this.matrixWorld);
+          this.matrixWorld = this.cachedMatrixWorld;
+        }
+      }
+    };
+
+    THREE.Object3D.prototype.applyMatrix = function(matrix) {
+      this.matrix.multiplyMatrices(matrix, this.matrix);
```
Comment (netpro2k) [inline src/systems/world-update.js]: may want to call the original function here instead to make it clear that these 2 lines are not part of the patch. Or at least block it out with comments.

```js
  THREE.Object3D.prototype.applyMatrix = (function(origApplyMatrix) {
    return function(matrix) {
      origApplyMatrix.call(this, matrix);
      this.matrixWorldNeedsUpdate = true;
      if (!this.matrixIsModified) {
        this.matrixIsModified = true;
        if (this.cachedMatrixWorld) {
          this.cachedMatrixWorld.copy(this.matrixWorld);
          this.matrixWorld = this.cachedMatrixWorld;
        }
      }
    }
  })(THREE.Object3D.prototype.applyMatrix)
```

File: src/systems/world-update.js [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+        this.updateMatrices();
+        this.matrixWorld.decompose(position, quaternion, target);
+
+        return target;
+      };
+    })();
+
+    THREE.Object3D.prototype.updateMatrix = function() {
+      this.matrix.compose(
```
Comment (netpro2k) [inline src/systems/world-update.js]: Like suggested for `applyMatrix` I would either call the original or block off the original with comments to show what the actual monkeypatch is.

File: src/systems/world-update.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+      const position = new THREE.Vector3();
+      const scale = new THREE.Vector3();
+
+      return function getWorldQuaternion(target) {
+        if (target === undefined) {
+          console.warn("THREE.Object3D: .getWorldQuaternion() target is now required");
+          target = new THREE.Quaternion();
+        }
+
+        this.updateMatrices();
```
Comment (netpro2k) [inline src/systems/world-update.js]: Same as above, I would note with a comment the monkeypatched `updateMatricies` line

File: src/systems/world-update.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      const position = new THREE.Vector3();
+      const quaternion = new THREE.Quaternion();
+
+      return function getWorldScale(target) {
+        if (target === undefined) {
+          console.warn("THREE.Object3D: .getWorldScale() target is now required");
+          target = new THREE.Vector3();
+        }
+
+        this.updateMatrices();
```
Comment (netpro2k) [inline src/systems/world-update.js]: Same as above, I would note with a comment the monkeypatched `updateMatricies` line

File: src/systems/world-update.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+  },
+
+  _patchThreeJS: function() {
+    THREE.Object3D.prototype.getWorldPosition = function(target) {
+      if (target === undefined) {
+        console.warn("THREE.Object3D: .getWorldPosition() target is now required");
+        target = new THREE.Vector3();
+      }
+
+      this.updateMatrices();
```
Comment (netpro2k) [inline src/systems/world-update.js]: Hmm no clean way to call the original here so I would note with a comment the monkeypatched `updateMatricies` line

File: src/systems/world-update.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+      } else if (this.matrixNeedsUpdate || this.matrixAutoUpdate || forceLocalUpdate) {
+        this.updateMatrix();
+        if (this.matrixNeedsUpdate) this.matrixNeedsUpdate = false;
+      }
+
+      if (!skipParents && this.parent) {
+        this.parent.updateMatrices(false, forceWorldUpdate, false);
+      }
+
+      if (this.matrixWorldNeedsUpdate || forceWorldUpdate) {
```
Comment (netpro2k) [inline src/systems/world-update.js]: Don't we also need to do this if any parent matrix was updated above?

File: src/components/cursor-controller.js [hunk 11]
```diff
@@ -129,9 +131,10 @@ AFRAME.registerComponent("cursor-controller", {
       }
       cursor.object3D.position.copy(cursorPose.position).addScaledVector(cursorPose.direction, this.distance);
       // The cursor will always be oriented towards the player about its Y axis, so objects held by the cursor will rotate towards the player.
-      camera.object3D.getWorldPosition(cameraPos);
+      getLastWorldPosition(camera.object3D, cameraPos);
       cameraPos.y = cursor.object3D.position.y;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/cursor-controller.js]: I see, so theoretically we should be transforming it into cursor space before calling lookAt. I'll leave it for now since that'll just add time to a hotspot

File: src/components/destroy-at-extreme-distances.js [hunk 12]
```diff
@@ -12,10 +14,12 @@ AFRAME.registerComponent("destroy-at-extreme-distances", {
     const pos = new THREE.Vector3();
     return function() {
       const { xMin, xMax, yMin, yMax, zMin, zMax } = this.data;
-      this.el.object3D.getWorldPosition(pos);
-      this.el.parentNode === this.el.sceneEl
-        ? pos.copy(this.el.object3D.position)
-        : this.el.object3D.getWorldPosition(pos);
+      getLastWorldPosition(this.el.object3D, pos);
+
+      if (this.el.parentNode === this.el.sceneEl) {
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/destroy-at-extreme-distances.js]: ah, `getLastWorldPosition` (and `getWorldPosition` for that matter) are now both cheap to call (esp the first one)

File: src/components/gltf-model-plus.js [hunk 13]
```diff
@@ -160,6 +161,7 @@ const inflateEntities = function(node, templates, isRoot, modelToWorldScale) {
     // preserved under the group (but not the node). Otherwise `clipArray` will be
     // `null` in `THREE.AnimationClip.findByName`.
     node.parent.animations = node.animations;
+    node.matrixAutoUpdate = true;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/gltf-model-plus.js]: you're right. for some reason I thought that the node here was a potential animation target, but i can see the GLTF loader handles setting `matrixAutoUpdate` on all the relevant entities that are animated.

File: src/components/ik-controller.js [hunk 14]
```diff
@@ -87,34 +87,42 @@ AFRAME.registerComponent("ik-controller", {
   update(oldData) {
     if (this.data.leftEye !== oldData.leftEye) {
       this.leftEye = this.el.object3D.getObjectByName(this.data.leftEye);
+      this.leftEye.matrixAutoUpdate = true;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/ik-controller.js]: I can try to do this, but my mental model for `matrixAutoUpdate` if different -- in general I have felt flipping it to true for things where we know there's going to be a tick somewhere probably moving it every frame. i think these joints are a good example, since if they're visible, they're tracked and/or are skinned/IK'd against tracked objects in the skeleton, so ultimately they're probably moving each frame.

i don't mind reconsidering the way we think about when/how to flip the flag. if its OK with you I'd like to do this in later PRs where we try to incrementally remove these. (the benefit here is not expected performance gain, but like you said reducing implicit behavior.) i think probably we should talk through how we want to think about `matrixAutoUpdate` as a team though.

[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/ik-controller.js]: as an example of something where `matrixAutoUpdate = true` is set and I'm on the fence about even given my current heuristic of "moves every frame" is the interactable objects, since when pinned/locked we're still computing it. in general, we could probably flip it off, but we'll need to go in and update all the various call sites where the pos/rot/scale are changed like networked-aframe and super hands.

at 30k/ft worth mentioning that a ton of updates are now avoided by a) skipping invisible objects and b) making it a whitelist. so the material performance impact of flipping this flag off in any of the places it's currently on I expect to be minimal.

File: src/components/media-loader.js [hunk 15]
```diff
@@ -62,6 +63,8 @@ AFRAME.registerComponent("media-loader", {
       });
       this.shapeAdded = true;
     }
+
+    mesh.matrixNeedsUpdate = true;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/media-loader.js]: I agree the flag flip is not strictly necessary, but I think in general the rule of thumb is:

- if the object is `matrixAutoUpdate`, do not worry about flipping dirty bits on it
- if it isn't, *always* flip `matrixNeedsUpdate` in any piece of code that modifies the pos/rot/scale

this does mean that multiple behaviors may be redundantly flipping the flag, but that's the reason it's a flag, and not a call to `updateMatrix` (and justifies introducing it.)

[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/media-loader.js]: (worth noting if it is a point of confusion, setting `matrixNeedsUpdate` has no effect or performance cost on computing the local matrix of children, but does cause the childrens' world matrices to be recomputed at the end of the frame)

File: src/components/media-loader.js [hunk 16]
```diff
@@ -88,6 +91,7 @@ AFRAME.registerComponent("media-loader", {
       this.loaderMixer = new THREE.AnimationMixer(mesh);
       this.loadingClip = this.loaderMixer.clipAction(loadingObject.animations[0]);
       this.loadingClip.play();
+      mesh.matrixAutoUpdate = true;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/media-loader.js]: i'm on the fence about this one. i lean towards thinking this is in the category of things where it's sane to set this flag -- similarly to how it's done in the GLTF loader.

i don't think it's an either or either. adding the suggested change to `AnimationMixer` is also a good idea.

File: src/components/position-at-box-shape-border.js [hunk 17]
```diff
@@ -139,6 +140,8 @@ AFRAME.registerComponent("position-at-box-shape-border", {
         this.target.scale.setScalar(finalScale);
       }
 
+      this.target.matrixNeedsUpdate = true;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/position-at-box-shape-border.js]: ah the early out above is on the first tick which will force a matrix update, but I will add the flag anyway so its less brittle

File: src/components/set-sounds-invisible.js [hunk 18]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Sets all non-positional audio children to not be visible so their world matrices are not updated.
+ */
+AFRAME.registerComponent("set-sounds-invisible", {
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/set-sounds-invisible.js]: i suspect we could run into trouble here because mainline three.js doesn't have the constraint that "non-visible objects don't have their matrices updated" -- so a sane patch to A-Frame would be "set visible to false in the sound component" which would break positional audio for us

File: src/components/set-sounds-invisible.js [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+  schema: {
+    withinDistance: { type: "number" }
+  },
+
+  init() {
+    for (const component of Object.values(this.el.components)) {
+      if (component.attrName !== "sound" && !component.attrName.startsWith("sound__")) continue;
+      if (component.data.positional) continue;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/set-sounds-invisible.js]: reflecting on it i like it better breaking these up instead of doing the ((x && y) || z). the first guard clause is filtering the components the second is filtering the positional ones

File: src/components/set-sounds-invisible.js [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+      component.pool.traverse(o => (o.visible = false));
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/set-sounds-invisible.js]: the reason I did it this way is I didn't want to assume anything about what happens to these "pooled" entities -- i didn't read the code but was considering that A-Frame may rip them out of the hierarchy to "depool" them which would result in the matrices updating

File: src/components/spot-light.js [hunk 21]
```diff
@@ -15,6 +15,7 @@ AFRAME.registerComponent("spot-light", {
     this.light.target.position.set(0, 0, 1);
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/spot-light.js]: nice catch

File: src/components/teleport-controls-matrix-auto-update.js [hunk 22]
```diff
@@ -0,0 +1,9 @@
+AFRAME.registerComponent("teleport-controls-matrix-auto-update", {
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/teleport-controls-matrix-auto-update.js]: yeah this is pretty awful i agree. the current teleporter controls also is quite gnarly. it took me way too long to get this to work

File: src/hub.html [hunk 23]
```diff
@@ -100,16 +100,16 @@
...
-                    <a-entity class="left-controller"></a-entity>
+                    <a-entity class="left-controller" matrix-auto-update></a-entity>
 
-                    <a-entity class="right-controller"></a-entity>
+                    <a-entity class="right-controller" matrix-auto-update></a-entity>
 
                     <a-entity class="model" gltf-model-plus="inflate: true">
                         <template data-name="RootScene">
-                            <a-entity ik-controller hand-pose__left hand-pose__right space-invader-mesh="meshName: Bot_Skinned"></a-entity>
+                            <a-entity ik-controller hand-pose__left hand-pose__right space-invader-mesh="meshName: Bot_Skinned" matrix-auto-update></a-entity>
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/hub.html]: yup. worth noting that there is no additional overhead to adding this flag to objects that are having their matrices every frame -- the main reason you wouldn't want this flag as you mentioned is:

- the object doesn't actually move most frames so you're losing the optimization

- (tbd) it's an anti-pattern to use the flag for reasons you've mentioned

File: src/hub.html [hunk 24]
```diff
@@ -400,9 +407,9 @@
                 <a-entity in-world-hud="haptic:#player-right-controller;raycaster:#player-right-controller;" rotation="30 0 0">
                     <a-rounded height="0.08" width="0.2" color="#000000" position="-0.30 0.125 0" radius="0.040" opacity="0.35" class="hud bg"></a-rounded>
                     <a-image scale="0.07 0.06 0.06" position="-0.23 0.165 0.001" src="#presence-count" material="alphaTest:0.1;"></a-image>
-                    <a-entity id="hud-presence-count" text="value:; width:1.1; align:center;" position="-0.155 0.165 0"></a-entity>
+                    <a-entity id="hud-presence-count" text="value:; width:1.1; align:center;" position="-0.155 0.165 0" matrix-auto-update="target: text;"></a-entity>
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/hub.html]: the text component updates the transform of the underlying mesh based upon the contents -- so if the contents change, the matrix needs to be updated.

this is a good example of where this flag is preventing an optimization -- since this text is "slowly changing" I would argue regardless of the other concerns that it would be better for the text component to set the flag when the text changes.

File: src/hub.js [hunk 25]
```diff
@@ -109,6 +112,7 @@ const store = window.APP.store;
 const qs = new URLSearchParams(location.search);
 const isMobile = AFRAME.utils.device.isMobile();
 
+THREE.Object3D.DefaultMatrixAutoUpdate = false;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/hub.js]: haha things were pretty messed up, I'm glad nobody had to see our poor app when I first flipped it (and the magic `autoUpdate` flag on the scene)

File: src/hub.js [hunk 26]
```diff
@@ -188,6 +193,8 @@ function setupLobbyCamera() {
     camera.object3D.position.set(cameraPos.x, 2.5, cameraPos.z);
   }
 
+  camera.object3D.updateMatrix();
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/hub.js]: Haha I think this was the first matrix i updated when i started this patch, before `matrixNeedsUpdate` existed and I thought this patch would be just as simple as calling `updateMatrix()` in a bunch of places -- lol

File: src/hub.js [hunk 27]
```diff
@@ -334,6 +341,7 @@ document.addEventListener("DOMContentLoaded", async () => {
   }
 
   const scene = document.querySelector("a-scene");
+  scene.object3D.autoUpdate = false;
```
Comment (gfodor) [inline src/hub.js]: ^^ note other major change. this results in killing the world matrix update loop in three.js, which allows us to run it ourselves in the aframe system.

(incidentally, I'm not sure if this is totally necessary now, but I like having this loop under our control atm.)

File: src/systems/camera-mirror.js [hunk 28]
```diff
@@ -80,11 +80,15 @@ AFRAME.registerSystem("camera-mirror", {
       if (playerHead) {
         tempHeadScale.copy(playerHead.scale);
         playerHead.scale.set(1, 1, 1);
+        playerHead.updateMatrices(true, true);
+        playerHead.updateMatrixWorld(true, true);
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/systems/camera-mirror.js]: `updateMatrices` walks up the tree, then `updateMatrixWorld` walks down it.

`updateMatrixWorld` is definitely a textbook case of the negative impact of a poorly named function. you'd assume by reading it that it updates the world matrix of the object you call it on as its primary operation. in fact: 

- it does not update the world matrix correctly, because it assumes the parent's local + world matrices up the chain are up-to-date (they may not be if the parent has had its pos/rot/scale modified in this frame)
 
- updates the local matrix too (but only if it's auto updating)

- updates the entire subtree's world matrices as well (but only if the world matrix was stale)

`updateMatrixWorld` makes sense for the outer "update all the dirty world matrices" tick loop that three.js does, but it should have never been a public API. it's used in various pieces of code assuming that it's how you update the world matrix of the object you call it on.

the new monkey patch exposes `updateMatrices` for this use-case as a public API you can use to update the local and/or world matrices on the object you call it on. it will have no effect on the children of the object, and will ensure that the necessary parents' matrices are up-to-date (unless you force it to not do this work.)

File: src/systems/world-update.js (hunk 6 above)
Comment (gfodor) [reply src/systems/world-update.js]: ah yep i had it that way before, because i thought I didn't need to update `matrixWorldNeedsUpdate`, but I do, so I'll revert

File: src/systems/world-update.js (hunk 10 above)
Comment (gfodor) [reply src/systems/world-update.js]: that sounds more correct

File: src/systems/world-update.js (hunk 3 above)
Comment (gfodor) [reply src/systems/world-update.js]: `updateMatrixWorld` is recursing downwards, so if we passed `skipParents = false` to `updateMatrices` we'd:

- break the semantics of `updateMatrixWorld` inherited from three.js where there are no side effects on parents

- end up redundantly visiting each node an additional log(N) times (I think?)

File: src/systems/world-update.js (hunk 1 above)
Comment (gfodor) [reply src/systems/world-update.js]: the reason I added this was because one of my original intentions was to use the frame counter here to reduce redundant matrix computations by different components all calling `updateMatrixWorld` intra-frame. this ended up not mattering. we could go back to `autoUpdate`, lets talk tomorrow. not sure if it's a good idea or not.

File: src/systems/world-update.js (hunk 2 above)
Comment (gfodor) [reply src/systems/world-update.js]: this was incidentally very very useful for debugging since I could measure the updateMatrixWorld cost as part of the main tick, but we can probably take this out if we want

File: src/components/gltf-model-plus.js [hunk 29]
```diff
@@ -143,6 +143,7 @@ const inflateEntities = function(node, templates, isRoot, modelToWorldScale) {
 
   node.matrixAutoUpdate = false;
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/gltf-model-plus.js]: (you'll note I special case the situation where on the first tick the object still has the identity transform -- in which case `updateMatrix()` is *not* called)

File: src/components/bone-visibility.js [hunk 30]
```diff
@@ -16,6 +16,8 @@ AFRAME.registerComponent("bone-visibility", {
       }
 
       this.lastVisible = visible;
+      this.el.object3D.updateMatrices(true, true);
+      this.el.object3D.updateMatrixWorld(true, true);
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/bone-visibility.js]: answered in discord but since its not visible, we won't update the matrices, which matters because then it will be mis-skinned. (this took me forever to actually fix given the hairy nature of getting the matrix update routines both correct and fast)

the visibilty changes happen rarely enough I think it'd be more confusing than its worth to special case it for the transition -- in general this just reads "ensure all the matrices are always correct coming out of this function if visibility changes"

**Comment by gfodor** [comment]
one side comment -- I am not all that nervous about this monkey patch wrt API breakages etc. for the following reasons:

- the biggest change is a new API flag `matrixNeedsUpdate`, which is also in PR from takahiro and most likely will have identical semantics

- `updateMatrixWorld` has mostly the same semantics as the mainline, except it skips non-visible objects and for identity matrices copies the world matrix by reference. The visibility thing is a common and relatively sane optimization that others have patched in too, given the fact that visible objects' entire subtrees are not rendered by three. The main issue with the visibility thing is that there are some objects, like bones and positional audio nodes, that are not things you can see (and so are never really "visible" in a rendering sense) but have changing transforms -- so that's a bit weird.

- `updateMatrices` is basically just a new, self contained function that basically is "what people probably think updateMatrixWorld does" -- we could pull off of the prototype if we wanted to. but it seems like a good idea to have it be part of the Object3D API.

- the additional monkeypatches don't really break any expected behaviors and are at best just faster versions of the public API (eg I don't think anyone calling `getWorldPosition` intends for all the children world matrices to update.)

File: src/components/ik-controller.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/ik-controller.js]: i looked into updating animation mixer, the proper place would be in `PropertyBinding` -- there are a few places however in `SetterByBindingTypeAndVersioning`

File: src/components/teleport-controls-matrix-auto-update.js [hunk 31]
```diff
@@ -0,0 +1,9 @@
...
+  tick: function() {
+    const teleportControls = this.el.components["teleport-controls"];
+    if (this.lastHitEntity !== teleportControls.hitEntity) {
+      document.querySelectorAll(".hitEntity").forEach(o => o.object3D.traverse(x => (x.matrixAutoUpdate = true)));
```
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/teleport-controls-matrix-auto-update.js]: ohh i figured it out. there are multiple teleport controls.

File: src/components/gltf-model-plus.js [hunk 32]
```diff
@@ -143,6 +143,8 @@ const inflateEntities = function(node, templates, isRoot, modelToWorldScale) {
...
   node.matrix.identity();
+  node.matrix.decompose(node.position, node.rotation, node.scale);
+  el.object3D.matrixNeedsUpdate = true;
```
Comment (gfodor) [inline src/components/gltf-model-plus.js]: added for sanity's sake

File: src/components/gltf-model-plus.js [hunk 33]
```diff
@@ -256,6 +258,8 @@ async function loadGLTF(src, contentType, preferredTechnique, onProgress) {
   const envMap = await CachedEnvMapTexture;
 
   gltf.scene.traverse(object => {
+    object.matrixAutoUpdate = false;
```
Comment (gfodor) [inline src/components/gltf-model-plus.js]: needed because the GLTF loader will set it

File: src/components/gltf-model-plus.js [hunk 34]
```diff
@@ -361,6 +365,8 @@ AFRAME.registerComponent("gltf-model-plus", {
         (this.inflatedEl = inflateEntities(this.model, this.templates, true, this.data.modelToWorldScale))
       ) {
         this.el.appendChild(this.inflatedEl);
+        this.el.object3D.matrixNeedsUpdate = true; // Not totally sure why this is necessary
```
Comment (gfodor) [inline src/components/gltf-model-plus.js]: /shrug on this one. some kind of weird condition where the GLTF model is loading in so the "first update matrix happens" thing doesn't work. i'm not going to worry about it for now

File: src/components/media-loader.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (gfodor) [reply src/components/media-loader.js]: animator now patched

**Comment by gfodor** [comment]
https://github.com/mozilla/hubs/pull/756
